common/djangoapps/student/models.py

```diff
@@ -331,6 +362,79 @@ def user_profile_pre_save_callback(sender, **kwargs):
     if user_profile.requires_parental_consent() and user_profile.has_profile_image:
         user_profile.profile_image_uploaded_at = None
 
+    # Cache "old" field values on the model instance so that they can be
+    # retrieved in the post_save callback when we emit an event with new and
+    # old field values.
+    user_profile._changed_fields = get_changed_fields_dict(user_profile, sender)
+
+@receiver(post_save, sender=UserProfile)
+def user_profile_post_save_callback(sender, **kwargs):
+    """
+    Emit analytics events after saving the UserProfile.
+    """
+    user_profile = kwargs['instance']
+    # pylint: disable=protected-access
+    emit_field_changed_events(
+        user_profile, user_profile.user, USER_SETTINGS_CHANGED_EVENT_NAME, sender._meta.db_table, ['meta']
+    )
+
+
+# @receiver(pre_save, sender=User)
+# def user_pre_save_callback(sender, **kwargs):
+#     """
+#     Capture old fields on the user instance before save and cache them as a
+#     private field on the current model for use in the post_save callback.
+#     """
+#     user = kwargs['instance']
+#     user._changed_fields = get_changed_fields_dict(user, sender)
+
+
+# @receiver(post_save, sender=User)
+# def user_post_save_callback(sender, **kwargs):
+#     """
+#     Emit analytics events after saving the User.
+#     """
+#     user = kwargs['instance']
+#     # pylint: disable=protected-access
+#     emit_field_changed_events(
+#         user, user, USER_SETTINGS_CHANGED_EVENT_NAME, sender._meta.db_table, hidden_fields=['password']
+#     )
+
+def emit_field_changed_events(instance, user, event_name, db_table, excluded_fields=None, hidden_fields=None):
+    """
+    For the given model instance, save the fields that have changed since the
+    last save.
+
+    Args:
+        instance (Model instance): the model instance that is being saved
+        user (User): the user that this instance is associated with
+        event_name (str): the name of the event to be emitted
+        db_table (str): the name of the table that we're modifying
+        excluded_fields (list): a list of field names for which events should
+            not be emitted
+        hidden_fields (list): a list of field names specifying fields whose
+            values should not be included in the event (None will be used
+            instead)
+
+    Returns:
+        None
+    """
+    excluded_fields = excluded_fields or []
+    hidden_fields = hidden_fields or []
+    changed_fields = getattr(instance, '_changed_fields', [])
+    for field in changed_fields:
```

Comment:
I'm wondering whether you can clear the changed fields property once you've sent the events. It seems like a memory leak to hold on to them.